signer/src/network/libp2p/swarm.rs

```diff
@@ -70,22 +85,66 @@ impl SignerBehavior {
 
         let bootstrap_config = bootstrap::Config::new(local_peer_id)
             .with_initial_delay(config.initial_bootstrap_delay)
-            .add_seed_addresses(config.seed_addresses);
-
+            .add_seed_addresses(&config.seed_addresses);
         let bootstrap = bootstrap::Behavior::new(bootstrap_config);
 
         Ok(Self {
             gossipsub: Self::gossipsub(&keypair)?,
             mdns,
-            kademlia: Self::kademlia(local_peer_id),
+            kademlia,
             ping: Default::default(),
             identify,
             autonat_client,
             autonat_server,
             bootstrap,
+            connection_limits: Self::connection_limits(&config),
         })
     }
 
+    fn connection_limits(config: &SignerSwarmConfig) -> connection_limits::Behaviour {
+        // The number of signers is the number of signers in the set, minus one
+        // for the local signer. This is used to calculate the connection limits.
+        // Note: We use `max(2)` here to ensure that we always have limits
+        // that support at least two signers.
+        let num_signers = config.num_signers.saturating_sub(1).max(2) as u32;
+
+        // Allow `num_signers` * 3 connections to be established at once. This
+        // allows room for one incoming/outgoing connection to each signer, plus
+        // one additional connection per signer as a buffer for protocols such
+        // as autonat.
+        let max_established = num_signers.saturating_mul(3);
```

Comment:
I'm wondering if we should drop this limit in favour of just using `max_established_per_peer`. Are we gaining something by having this explicit one as well as the per peer one? Maybe also drop `max_established_incoming` for the same reason

Reply:
I think that having a general upper-bound on our established connections when we know roughly how many connections we should be expecting (at a maximum) is generally a good practice. If this isn't limited, then resource exhaustion attacks are theoretically possible as we're letting the connections "go further" than they otherwise would.

Why I'm skeptical to remove this (and `max_established_incoming`) is because the documentation explicitly mentions these two parameters for controlling eclipse attacks, where `max_established_per_peer` is `inbound+outbound`, this lets us explicitly control our inbound connection ratio, making me think we should probably even reduce `max_established_incoming`.